### PR TITLE
feat: approve command

### PR DESCRIPTION
**Pull request by raffis**
# What does this change
Hey me again 🙈 ,

This pr adds the ability for approving pull requests supported by all platforms.
Basically reviewers now get an easy option to approve pr's made by multi-gitter.
Usually pull requests require approval in branch protected repos and there is currently no good way to batch approve multi pr's.
For example we collect all batch operations in a repository, a reviewer can grab the config from there and start a straight forward review process with the same tool. (Using the --batch option this could even be fully automated in a ci pipeline).

This pr adds the `review` command with a couple of options:
* --comment for a review comment (can also be overwritten while reviewing pr's)
* --all to batch review all of them. Basically review a diff output from all pr's. Useful if you have 100 pr's with a single line change where you don't want to review each individually.
* --batch=x Approve/Decline/Comment all targeted pull requests without asking questions
* --no-pager This pr pipes the diffs to a pager (less, more (windows), PAGER env). If no pager should be used one can dump to stdout directly using this option
* --include-approved By default approve ignores pull requests which are already approved by the reviewing person.

# What issue does it fix
Currently reviewing is hard. If you open 100 pr's with a simple change you either have to wait until people reviewed them, try to push them or whatever... There are also some workarounds like verify 2 and then batch approve everything via the api.

# Notes for the reviewer
* No tests added so far, waiting for your comments if this will be merged.
* I did not test it on windows yet
* I was unable to test bitbucket so far
* The autogenerated bitbucket go api seems a bit messy. The pr comment api is not there even though its documented.  Alternative bitbucket implementations do not look any better.

# Checklist
- [x] Made sure the PR follows the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines
- [ ] Tests if something new is added

### CODE REVIEW

**Comment by lindell** [comment]
Hey. Thanks for the PR! But please create an issue discussion about the addition so that you don't create anything unnecessarily ;) 

I do think this is a good addition but will have to take some time before looking at it since the changes are extensive. Let me get back to you with a review in a few days :)

**Comment by raffis** [comment]
> Hey. Thanks for the PR! But please create an issue discussion about the addition so that you don't create anything unnecessarily ;)
> 
> I do think this is a good addition but will have to take some time before looking at it since the changes are extensive. Let me get back to you with a review in a few days :)

Yeah should have commented (fixes #295). Sure take your time 👍🏻

File: internal/multigitter/review.go [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		}
+
+		approvedPrs = append(approvedPrs, approvdPr{
+			PullRequest: pr,
+			approved:    approved,
+		})
+	}
+
+	if len(approvedPrs) == 0 {
+		fmt.Println("All pull requests are approved by you. You can still view and comment them by using review --include-approved")
```
Comment (lindell) [inline internal/multigitter/review.go:68]: Should probably have a text for `len(prs) == 0`, so that you don't get that you have all approved PRs when none exist, or you misspelled the branch-name 😉

File: internal/multigitter/review.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+func (s Reviewer) Review(ctx context.Context) error {
+	prs, err := s.VersionController.GetPullRequests(ctx, s.FeatureBranch)
+	if err != nil {
+		return err
+	}
+
+	var approvedPrs []approvdPr
+
+	for _, pr := range prs {
+		approved, err := s.VersionController.IsPullRequestApprovedByMe(ctx, pr)
```
Comment (lindell) [inline internal/multigitter/review.go:51]:
```suggestion
	for _, pr := range prs {
		log := log.WithField("pr", pr.String())
		approved, err := s.VersionController.IsPullRequestApprovedByMe(ctx, pr)
```

Comment (lindell) [reply internal/multigitter/review.go:51]: To be able to get information of which PR went wrong when logging, we need to include the PR as a part of the logging.

File: internal/multigitter/review.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+	if len(approvedPrs) == 0 {
+		fmt.Println("All pull requests are approved by you. You can still view and comment them by using review --include-approved")
+		return nil
+	}
+
+	var reviewDiffs string
+
+	for _, pr := range approvedPrs {
+		diff, err := s.VersionController.DiffPullRequest(ctx, pr.PullRequest)
```
Comment (lindell) [inline internal/multigitter/review.go:75]:
```suggestion
	for _, pr := range approvedPrs {
		log := log.WithField("pr", pr.String())
		
		diff, err := s.VersionController.DiffPullRequest(ctx, pr.PullRequest)
```

File: cmd/cmd-review.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	return cmd
+}
+
+func review(cmd *cobra.Command, _ []string) error {
+	flag := cmd.Flags()
+
+	branchName, _ := flag.GetString("branch")
+	comment, _ := flag.GetString("comment")
+	all, _ := flag.GetBool("all")
+	batch, _ := flag.GetString("batch")
```
[the remark this answers is not in the document]
Comment (lindell) [reply cmd/cmd-review.go:43]: Whatever it is called, it should be converted to an "enum" here, and therefore verified to be correct, before calling out to the multigitter package. 

See conflict strategy as an example

**Review by lindell — COMMENTED** [review]
Some initial comments. I've not reviewed all files yet.

File: internal/multigitter/review.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+	}
+
+	if s.All || s.Batch != "" {
+		return s.leaveReviews(ctx, strings.NewReader(reviewDiffs), approvedPrs...)
+	}
+
+	return nil
+}
+
+func (s Reviewer) printDiff(r io.ReadSeeker) error {
```
Comment (lindell) [inline internal/multigitter/review.go:97]: Why is a Reader(Seeker) used here and other functions that are called from it, instead of a string? It seems that it is always created from a string, so there is no benefit of streaming the data?

File: internal/multigitter/review.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+			log.Errorf("Error occurred while retrieving diff: %s", err.Error())
+			continue
+		}
+
+		if !s.All && s.Batch == "" {
+			if err := s.leaveReview(ctx, strings.NewReader(fmt.Sprintf("%s\n# %s\n%s\n%s", header, pr.String(), header, diff)), pr); err != nil {
+				log.Errorf("Error occurred while reviewing pr: %s", err.Error())
+			}
+		} else {
+			reviewDiffs = fmt.Sprintf("%s\n%s\n# %s\n%s\n%s", reviewDiffs, header, pr.String(), header, diff)
```
Comment (lindell) [inline internal/multigitter/review.go:86]: With large diffs, this might be slow, since a new string has to be recreated, and data copied over for each new PR. An example fix can be seen in the code suggestion

File: internal/multigitter/review.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+		}
+	}
+
+	if s.All || s.Batch != "" {
+		return s.leaveReviews(ctx, strings.NewReader(reviewDiffs), approvedPrs...)
+	}
```
Comment (lindell) [inline internal/multigitter/review.go:92]: A suggested improvement. The assumption is that All and Batch will be mutually inclusive. 
Removes some duplication of the Sprintf, and will avoid slow additions to reviewDiffs for large diffs.

```suggestion
	var reviewDiffs strings.Builder

	for _, pr := range approvedPrs {
		diff, err := s.VersionController.DiffPullRequest(ctx, pr.PullRequest)
		if err != nil {
			log.Errorf("Error occurred while retrieving diff: %s", err.Error())
			continue
		}

		reviewDiff := fmt.Sprintf("%s\n# %s\n%s\n%s", header, pr.String(), header, diff)

		if !s.All {
			if err := s.leaveReview(ctx, reviewDiff, pr); err != nil {
				log.Errorf("Error occurred while reviewing pr: %s", err.Error())
			}
		} else {
			reviewDiffs.WriteString("\n")
			reviewDiffs.WriteString(reviewDiff)
		}
	}

	if s.All {
		return s.leaveReviews(ctx, reviewDiffs.String(), approvedPrs...)
	}
```

File: internal/multigitter/review.go (hunk 1 above)
Comment (raffis) [reply internal/multigitter/review.go:69]: Yep :) 👍🏻

File: cmd/cmd-review.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (raffis) [reply cmd/cmd-review.go:43]: Batch always refers to all pulls, so it includes --all not matter if a user would set it or not (It's describe like that in the flag helper). While you can do all without the batch option. If all is set without batch a user gets the diff for all pull requests at once.

[the remark this answers is not in the document]
Comment (raffis) [reply cmd/cmd-review.go:43]: Agreed with the enum, definitely makes sense 👍🏻
 
With the flag name, from a user perspective and cli tooling I would pick batch as its more obvious that something is going on which does not require user interaction which is not the case if it's called state. It could be --batch and --state=x. But thats an additional flag for no reason.

File: internal/multigitter/review.go (hunk 5 above)
Comment (raffis) [reply internal/multigitter/review.go]: No reason anymore, was required before I changed some bits. Changed to string now.

**Comment by raffis** [comment]
Just used this new command to batch review and merge hundred pr's opened by renovate (not created using multi-gitter itself). Thats also a nice use case for this command.

File: internal/multigitter/review.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+			}
+		} else {
+			reviewDiffs.WriteString("\n")
+			reviewDiffs.WriteString(reviewDiff)
+		}
+	}
+
+	if s.All || s.BatchOperation != nil {
+		return s.leaveReviews(ctx, reviewDiffs.String(), reviewPrs...)
+	}
```
Comment (lindell) [inline internal/multigitter/review.go:98]: The logic for the Batch/All flow and the manual review code is spread out inside the loop, and after it. Making it hard to follow what is happening for the different ways the command can work. The same can be said for `leaveReviews` where it is both code for "Batch" and "All" while these could be separate functions to make it easier to understand what is going on in each individual case. 

In general, it's better to break out the logic that is reused into different functions, and call those it in the different "flows", instead of putting multiple flows into one function where the shared code will live. Since the reader have to parse all logic at once, instead of being able to look at each part individually.

As a POC of this, I created https://github.com/lindell/multi-gitter/commit/277408a58ae43e07e38f2ce23fa09a1e4c3c6641. I tried to limit the changes to just this concept, so there are things like how `leaveReviews` calls into `reviewAction` that should be discussed, but I leave that to a later discussion.

Thoughts? 🙂

Comment (raffis) [reply internal/multigitter/review.go]: Merged picked up your commit, fine with the changes, just wanna get this released.

File: internal/multigitter/multigitter.go [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+type VersionController interface {
+	GetRepositories(ctx context.Context) ([]scm.Repository, error)
+	CreatePullRequest(ctx context.Context, repo scm.Repository, prRepo scm.Repository, newPR scm.NewPullRequest) (scm.PullRequest, error)
+	GetPullRequests(ctx context.Context, branchName string) ([]scm.PullRequest, error)
+	GetOpenPullRequest(ctx context.Context, repo scm.Repository, branchName string) (scm.PullRequest, error)
+	MergePullRequest(ctx context.Context, pr scm.PullRequest) error
+	IsPullRequestApprovedByMe(ctx context.Context, pullReq scm.PullRequest) (bool, error)
+	ApprovePullRequest(ctx context.Context, pr scm.PullRequest, comment string) error
+	RejectPullRequest(ctx context.Context, pr scm.PullRequest, comment string) error
+	CommentPullRequest(ctx context.Context, pr scm.PullRequest, comment string) error
```
[the remark this answers is not in the document]
Comment (raffis) [reply internal/multigitter/multigitter.go]: Done

File: cmd/cmd-review.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (raffis) [reply cmd/cmd-review.go:47]: I think --batch is more common to understand as a user that no additional follow up questions are required and it can be used silently. Might also call it silent instead.
If batch is just ditched and its called state I think it's not exactly what it's intended for.

**Comment by fat0** [comment]
Hi, would really really appreciate this feature being added please.
